File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebase/auth/FlutterFirebaseAuthUser.java
/*
 * Copyright 2023, the Chromium project authors.  Please see the AUTHORS file
 * for details. All rights reserved. Use of this source code is governed by a
 * BSD-style license that can be found in the LICENSE file.
 */

package io.flutter.plugins.firebase.auth;

import static io.flutter.plugins.firebase.core.FlutterFirebasePlugin.cachedThreadPool;

import android.app.Activity;
import android.net.Uri;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.FirebaseApp;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GetTokenResult;
import com.google.firebase.auth.OAuthProvider;
import com.google.firebase.auth.PhoneAuthCredential;
import com.google.firebase.auth.UserProfileChangeRequest;
import java.util.Map;

public class FlutterFirebaseAuthUser
    implements GeneratedAndroidFirebaseAuth.FirebaseAuthUserHostApi {

  private Activity activity;

  public void setActivity(Activity activity) {
    this.activity = activity;
  }

  public static FirebaseUser getCurrentUserFromPigeon(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp pigeonApp) {
    FirebaseApp app = FirebaseApp.getInstance(pigeonApp.getAppName());
    FirebaseAuth auth = FirebaseAuth.getInstance(app);
    if (pigeonApp.getTenantId() != null) {
      auth.setTenantId(pigeonApp.getTenantId());
    }

    return auth.getCurrentUser();
  }

  @Override
  public void delete(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @NonNull GeneratedAndroidFirebaseAuth.VoidResult result) {
    FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);

    if (firebaseUser == null) {
      result.error(FlutterFirebaseAuthPluginException.noUser());
      return;
    }

    firebaseUser
        .delete()
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                result.success();
              } else {
                result.error(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        task.getException()));
              }
            });
  }

  @Override
  public void getIdToken(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @NonNull Boolean forceRefresh,
      @NonNull
          GeneratedAndroidFirebaseAuth.Result<GeneratedAndroidFirebaseAuth.PigeonIdTokenResult>
              result) {
    cachedThreadPool.execute(
        () -> {
          FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);

          if (firebaseUser == null) {
            result.error(FlutterFirebaseAuthPluginException.noUser());
            return;
          }
          try {
            GetTokenResult response = Tasks.await(firebaseUser.getIdToken(forceRefresh));
            result.success(PigeonParser.parseTokenResult(response));
          } catch (Exception exception) {
            result.error(FlutterFirebaseAuthPluginException.parserExceptionToFlutter(exception));
          }
        });
  }

  @Override
  public void linkWithCredential(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @NonNull Map<String, Object> input,
      @NonNull
          GeneratedAndroidFirebaseAuth.Result<GeneratedAndroidFirebaseAuth.PigeonUserCredential>
              result) {
    FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);
    AuthCredential credential = PigeonParser.getCredential(input);

    if (firebaseUser == null) {
      result.error(FlutterFirebaseAuthPluginException.noUser());
      return;
    }

    if (credential == null) {
      result.error(FlutterFirebaseAuthPluginException.invalidCredential());
      return;
    }

    firebaseUser
        .linkWithCredential(credential)
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                result.success(PigeonParser.parseAuthResult(task.getResult()));
              } else {
                result.error(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        task.getException()));
              }
            });
  }

  @Override
  public void linkWithProvider(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @NonNull GeneratedAndroidFirebaseAuth.PigeonSignInProvider signInProvider,
      @NonNull
          GeneratedAndroidFirebaseAuth.Result<GeneratedAndroidFirebaseAuth.PigeonUserCredential>
              result) {
    FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);

    OAuthProvider.Builder provider = OAuthProvider.newBuilder(signInProvider.getProviderId());
    if (signInProvider.getScopes() != null) {
      provider.setScopes(signInProvider.getScopes());
    }
    if (signInProvider.getCustomParameters() != null) {
      provider.addCustomParameters(signInProvider.getCustomParameters());
    }

    firebaseUser
        .startActivityForLinkWithProvider(activity, provider.build())
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                result.success(PigeonParser.parseAuthResult(task.getResult()));
              } else {
                result.error(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        task.getException()));
              }
            });
  }

  @Override
  public void reauthenticateWithCredential(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @NonNull Map<String, Object> input,
      @NonNull
          GeneratedAndroidFirebaseAuth.Result<GeneratedAndroidFirebaseAuth.PigeonUserCredential>
              result) {
    FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);
    AuthCredential credential = PigeonParser.getCredential(input);

    if (firebaseUser == null) {
      result.error(FlutterFirebaseAuthPluginException.noUser());
      return;
    }

    if (credential == null) {
      result.error(FlutterFirebaseAuthPluginException.invalidCredential());
      return;
    }

    firebaseUser
        .reauthenticateAndRetrieveData(credential)
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                result.success(PigeonParser.parseAuthResult(task.getResult()));
              } else {
                result.error(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        task.getException()));
              }
            });
  }

  @Override
  public void reauthenticateWithProvider(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @NonNull GeneratedAndroidFirebaseAuth.PigeonSignInProvider signInProvider,
      @NonNull
          GeneratedAndroidFirebaseAuth.Result<GeneratedAndroidFirebaseAuth.PigeonUserCredential>
              result) {
    FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);

    OAuthProvider.Builder provider = OAuthProvider.newBuilder(signInProvider.getProviderId());
    if (signInProvider.getScopes() != null) {
      provider.setScopes(signInProvider.getScopes());
    }
    if (signInProvider.getCustomParameters() != null) {
      provider.addCustomParameters(signInProvider.getCustomParameters());
    }

    firebaseUser
        .startActivityForReauthenticateWithProvider(activity, provider.build())
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                result.success(PigeonParser.parseAuthResult(task.getResult()));
              } else {
                result.error(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        task.getException()));
              }
            });
  }

  @Override
  public void reload(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @NonNull
          GeneratedAndroidFirebaseAuth.Result<GeneratedAndroidFirebaseAuth.PigeonUserDetails>
              result) {
    FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);

    if (firebaseUser == null) {
      result.error(FlutterFirebaseAuthPluginException.noUser());
      return;
    }

    firebaseUser
        .reload()
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                result.success(PigeonParser.parseFirebaseUser(firebaseUser));
              } else {
                result.error(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        task.getException()));
              }
            });
  }

  @Override
  public void sendEmailVerification(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @Nullable GeneratedAndroidFirebaseAuth.PigeonActionCodeSettings actionCodeSettings,
      @NonNull GeneratedAndroidFirebaseAuth.VoidResult result) {
    FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);

    if (firebaseUser == null) {
      result.error(FlutterFirebaseAuthPluginException.noUser());
      return;
    }

    if (actionCodeSettings == null) {
      firebaseUser
          .sendEmailVerification()
          .addOnCompleteListener(
              task -> {
                if (task.isSuccessful()) {
                  result.success();
                } else {
                  result.error(
                      FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                          task.getException()));
                }
              });
      return;
    }

    firebaseUser
        .sendEmailVerification(PigeonParser.getActionCodeSettings(actionCodeSettings))
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                result.success();
              } else {
                result.error(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        task.getException()));
              }
            });
  }

  @Override
  public void unlink(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @NonNull String providerId,
      @NonNull
          GeneratedAndroidFirebaseAuth.Result<GeneratedAndroidFirebaseAuth.PigeonUserCredential>
              result) {
    FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);

    if (firebaseUser == null) {
      result.error(FlutterFirebaseAuthPluginException.noUser());
      return;
    }

    firebaseUser
        .unlink(providerId)
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                result.success(PigeonParser.parseAuthResult(task.getResult()));
              } else {
                Exception exception = task.getException();
                if (exception
                    .getMessage()
                    .contains("User was not linked to an account with the given provider.")) {
                  result.error(FlutterFirebaseAuthPluginException.noSuchProvider());
                } else {
                  result.error(
                      FlutterFirebaseAuthPluginException.parserExceptionToFlutter(exception));
                }
              }
            });
  }

  @Override
  public void updateEmail(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @NonNull String newEmail,
      @NonNull
          GeneratedAndroidFirebaseAuth.Result<GeneratedAndroidFirebaseAuth.PigeonUserDetails>
              result) {
    FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);

    if (firebaseUser == null) {
      result.error(FlutterFirebaseAuthPluginException.noUser());
      return;
    }

    firebaseUser
        .updateEmail(newEmail)
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                firebaseUser
                    .reload()
                    .addOnCompleteListener(
                        reloadTask -> {
                          if (reloadTask.isSuccessful()) {
                            result.success(PigeonParser.parseFirebaseUser(firebaseUser));
                          } else {
                            result.error(
                                FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                                    reloadTask.getException()));
                          }
                        });
              } else {
                result.error(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        task.getException()));
              }
            });
  }

  @Override
  public void updatePassword(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @NonNull String newPassword,
      @NonNull
          GeneratedAndroidFirebaseAuth.Result<GeneratedAndroidFirebaseAuth.PigeonUserDetails>
              result) {
    FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);

    if (firebaseUser == null) {
      result.error(FlutterFirebaseAuthPluginException.noUser());
      return;
    }

    firebaseUser
        .updatePassword(newPassword)
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                firebaseUser
                    .reload()
                    .addOnCompleteListener(
                        reloadTask -> {
                          if (reloadTask.isSuccessful()) {
                            result.success(PigeonParser.parseFirebaseUser(firebaseUser));
                          } else {
                            result.error(
                                FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                                    reloadTask.getException()));
                          }
                        });
              } else {
                result.error(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        task.getException()));
              }
            });
  }

  @Override
  public void updatePhoneNumber(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @NonNull Map<String, Object> input,
      @NonNull
          GeneratedAndroidFirebaseAuth.Result<GeneratedAndroidFirebaseAuth.PigeonUserDetails>
              result) {
    FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);

    if (firebaseUser == null) {
      result.error(FlutterFirebaseAuthPluginException.noUser());
      return;
    }

    PhoneAuthCredential phoneAuthCredential =
        (PhoneAuthCredential) PigeonParser.getCredential(input);

    if (phoneAuthCredential == null) {
      result.error(FlutterFirebaseAuthPluginException.invalidCredential());
      return;
    }

    firebaseUser
        .updatePhoneNumber(phoneAuthCredential)
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                firebaseUser
                    .reload()
                    .addOnCompleteListener(
                        reloadTask -> {
                          if (reloadTask.isSuccessful()) {
                            result.success(PigeonParser.parseFirebaseUser(firebaseUser));
                          } else {
                            result.error(
                                FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                                    reloadTask.getException()));
                          }
                        });
              } else {
                result.error(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        task.getException()));
              }
            });
  }

  @Override
  public void updateProfile(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @NonNull GeneratedAndroidFirebaseAuth.PigeonUserProfile profile,
      @NonNull
          GeneratedAndroidFirebaseAuth.Result<GeneratedAndroidFirebaseAuth.PigeonUserDetails>
              result) {
    FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);

    if (firebaseUser == null) {
      result.error(FlutterFirebaseAuthPluginException.noUser());
      return;
    }

    UserProfileChangeRequest.Builder builder = new UserProfileChangeRequest.Builder();

    if (profile.getDisplayNameChanged()) {
      builder.setDisplayName(profile.getDisplayName());
    }

    if (profile.getPhotoUrlChanged()) {
      if (profile.getPhotoUrl() != null) {
        builder.setPhotoUri(Uri.parse(profile.getPhotoUrl()));
      } else {
        builder.setPhotoUri(null);
      }
    }

    firebaseUser
        .updateProfile(builder.build())
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                firebaseUser
                    .reload()
                    .addOnCompleteListener(
                        reloadTask -> {
                          if (reloadTask.isSuccessful()) {
                            result.success(PigeonParser.parseFirebaseUser(firebaseUser));
                          } else {
                            result.error(
                                FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                                    reloadTask.getException()));
                          }
                        });
              } else {
                result.error(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        task.getException()));
              }
            });
  }

  @Override
  public void verifyBeforeUpdateEmail(
      @NonNull GeneratedAndroidFirebaseAuth.AuthPigeonFirebaseApp app,
      @NonNull String newEmail,
      @Nullable GeneratedAndroidFirebaseAuth.PigeonActionCodeSettings actionCodeSettings,
      @NonNull GeneratedAndroidFirebaseAuth.VoidResult result) {
    FirebaseUser firebaseUser = getCurrentUserFromPigeon(app);

    if (firebaseUser == null) {
      result.error(FlutterFirebaseAuthPluginException.noUser());
      return;
    }

    if (actionCodeSettings == null) {
      firebaseUser
          .verifyBeforeUpdateEmail(newEmail)
          .addOnCompleteListener(
              task -> {
                if (task.isSuccessful()) {
                  result.success();
                } else {
                  result.error(
                      FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                          task.getException()));
                }
              });
      return;
    }

    firebaseUser
        .verifyBeforeUpdateEmail(newEmail, PigeonParser.getActionCodeSettings(actionCodeSettings))
        .addOnCompleteListener(
            task -> {
              if (task.isSuccessful()) {
                result.success();
              } else {
                result.error(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        task.getException()));
              }
            });
  }
}

File: packages/firebase_auth/firebase_auth/android/src/main/kotlin/io/flutter/plugins/firebase/auth/FlutterFirebaseAuthUser.kt
/*
 * Copyright 2023, the Chromium project authors.  Please see the AUTHORS file
 * for details. All rights reserved. Use of this source code is governed by a
 * BSD-style license that can be found in the LICENSE file.
 */
package io.flutter.plugins.firebase.auth

import android.app.Activity
import android.net.Uri
import com.google.android.gms.tasks.Tasks
import com.google.firebase.FirebaseApp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.OAuthProvider
import com.google.firebase.auth.PhoneAuthCredential
import com.google.firebase.auth.UserProfileChangeRequest
import io.flutter.plugins.firebase.core.FlutterFirebasePlugin.cachedThreadPool

class FlutterFirebaseAuthUser : FirebaseAuthUserHostApi {
  private var activity: Activity? = null

  fun setActivity(activity: Activity?) {
    this.activity = activity
  }

  override fun delete(app: AuthPigeonFirebaseApp, callback: (Result<Unit>) -> Unit) {
    val firebaseUser = getCurrentUserFromPigeon(app)
    if (firebaseUser == null) {
      callback(Result.failure(FlutterFirebaseAuthPluginException.noUser()))
      return
    }

    firebaseUser.delete().addOnCompleteListener { task ->
      if (task.isSuccessful) {
        callback(Result.success(Unit))
      } else {
        callback(
            Result.failure(
                FlutterFirebaseAuthPluginException.parserExceptionToFlutter(task.exception)))
      }
    }
  }

  override fun getIdToken(
      app: AuthPigeonFirebaseApp,
      forceRefresh: Boolean,
      callback: (Result<InternalIdTokenResult>) -> Unit
  ) {
    cachedThreadPool.execute {
      val firebaseUser = getCurrentUserFromPigeon(app)
      if (firebaseUser == null) {
        callback(Result.failure(FlutterFirebaseAuthPluginException.noUser()))
        return@execute
      }
      try {
        val response = Tasks.await(firebaseUser.getIdToken(forceRefresh))
        callback(Result.success(PigeonParser.parseTokenResult(response)))
      } catch (exception: Exception) {
        callback(
            Result.failure(FlutterFirebaseAuthPluginException.parserExceptionToFlutter(exception)))
      }
    }
  }

  override fun linkWithCredential(
      app: AuthPigeonFirebaseApp,
      input: Map<String?, Any?>,
      callback: (Result<InternalUserCredential>) -> Unit
  ) {
    val firebaseUser = getCurrentUserFromPigeon(app)
    val credential = PigeonParser.getCredential(input)

    if (firebaseUser == null) {
      callback(Result.failure(FlutterFirebaseAuthPluginException.noUser()))
      return
    }

    if (credential == null) {
      callback(Result.failure(FlutterFirebaseAuthPluginException.invalidCredential()))
      return
    }

    firebaseUser.linkWithCredential(credential).addOnCompleteListener { task ->
      if (task.isSuccessful) {
        callback(Result.success(PigeonParser.parseAuthResult(task.result)))
      } else {
        callback(
            Result.failure(
                FlutterFirebaseAuthPluginException.parserExceptionToFlutter(task.exception)))
      }
    }
  }

  override fun linkWithProvider(
      app: AuthPigeonFirebaseApp,
      signInProvider: InternalSignInProvider,
      callback: (Result<InternalUserCredential>) -> Unit
  ) {
    val firebaseUser = getCurrentUserFromPigeon(app)
    val provider = buildOAuthProvider(signInProvider)

    firebaseUser!!
        .startActivityForLinkWithProvider(checkNotNull(activity), provider)
        .addOnCompleteListener { task ->
          if (task.isSuccessful) {
            callback(Result.success(PigeonParser.parseAuthResult(task.result)))
          } else {
            callback(
                Result.failure(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(task.exception)))
          }
        }
  }

  override fun reauthenticateWithCredential(
      app: AuthPigeonFirebaseApp,
      input: Map<String?, Any?>,
      callback: (Result<InternalUserCredential>) -> Unit
  ) {
    val firebaseUser = getCurrentUserFromPigeon(app)
    val credential = PigeonParser.getCredential(input)

    if (firebaseUser == null) {
      callback(Result.failure(FlutterFirebaseAuthPluginException.noUser()))
      return
    }

    if (credential == null) {
      callback(Result.failure(FlutterFirebaseAuthPluginException.invalidCredential()))
      return
    }

    firebaseUser.reauthenticateAndRetrieveData(credential).addOnCompleteListener { task ->
      if (task.isSuccessful) {
        callback(Result.success(PigeonParser.parseAuthResult(task.result)))
      } else {
        callback(
            Result.failure(
                FlutterFirebaseAuthPluginException.parserExceptionToFlutter(task.exception)))
      }
    }
  }

  override fun reauthenticateWithProvider(
      app: AuthPigeonFirebaseApp,
      signInProvider: InternalSignInProvider,
      callback: (Result<InternalUserCredential>) -> Unit
  ) {
    val firebaseUser = getCurrentUserFromPigeon(app)
    val provider = buildOAuthProvider(signInProvider)

    firebaseUser!!
        .startActivityForReauthenticateWithProvider(checkNotNull(activity), provider)
        .addOnCompleteListener { task ->
          if (task.isSuccessful) {
            callback(Result.success(PigeonParser.parseAuthResult(task.result)))
          } else {
            callback(
                Result.failure(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(task.exception)))
          }
        }
  }

  override fun reload(app: AuthPigeonFirebaseApp, callback: (Result<InternalUserDetails>) -> Unit) {
    val firebaseUser = getCurrentUserFromPigeon(app)
    if (firebaseUser == null) {
      callback(Result.failure(FlutterFirebaseAuthPluginException.noUser()))
      return
    }

    firebaseUser.reload().addOnCompleteListener { task ->
      if (task.isSuccessful) {
        callback(Result.success(PigeonParser.parseFirebaseUser(firebaseUser)!!))
      } else {
        callback(
            Result.failure(
                FlutterFirebaseAuthPluginException.parserExceptionToFlutter(task.exception)))
      }
    }
  }

  override fun sendEmailVerification(
      app: AuthPigeonFirebaseApp,
      actionCodeSettings: InternalActionCodeSettings?,
      callback: (Result<Unit>) -> Unit
  ) {
    val firebaseUser = getCurrentUserFromPigeon(app)
    if (firebaseUser == null) {
      callback(Result.failure(FlutterFirebaseAuthPluginException.noUser()))
      return
    }

    val task =
        if (actionCodeSettings == null) {
          firebaseUser.sendEmailVerification()
        } else {
          firebaseUser.sendEmailVerification(PigeonParser.getActionCodeSettings(actionCodeSettings))
        }

    task.addOnCompleteListener { completed ->
      if (completed.isSuccessful) {
        callback(Result.success(Unit))
      } else {
        callback(
            Result.failure(
                FlutterFirebaseAuthPluginException.parserExceptionToFlutter(completed.exception)))
      }
    }
  }

  override fun unlink(
      app: AuthPigeonFirebaseApp,
      providerId: String,
      callback: (Result<InternalUserCredential>) -> Unit
  ) {
    val firebaseUser = getCurrentUserFromPigeon(app)
    if (firebaseUser == null) {
      callback(Result.failure(FlutterFirebaseAuthPluginException.noUser()))
      return
    }

    firebaseUser.unlink(providerId).addOnCompleteListener { task ->
      if (task.isSuccessful) {
        callback(Result.success(PigeonParser.parseAuthResult(task.result)))
      } else {
        val exception = task.exception
        if (exception
            ?.message
            ?.contains("User was not linked to an account with the given provider.") == true) {
          callback(Result.failure(FlutterFirebaseAuthPluginException.noSuchProvider()))
        } else {
          callback(
              Result.failure(
                  FlutterFirebaseAuthPluginException.parserExceptionToFlutter(exception)))
        }
      }
    }
  }

  override fun updateEmail(
      app: AuthPigeonFirebaseApp,
      newEmail: String,
      callback: (Result<InternalUserDetails>) -> Unit
  ) {
    reloadAfterUserUpdate(getCurrentUserFromPigeon(app), callback) { it.updateEmail(newEmail) }
  }

  override fun updatePassword(
      app: AuthPigeonFirebaseApp,
      newPassword: String,
      callback: (Result<InternalUserDetails>) -> Unit
  ) {
    reloadAfterUserUpdate(getCurrentUserFromPigeon(app), callback) {
      it.updatePassword(newPassword)
    }
  }

  override fun updatePhoneNumber(
      app: AuthPigeonFirebaseApp,
      input: Map<String?, Any?>,
      callback: (Result<InternalUserDetails>) -> Unit
  ) {
    val firebaseUser = getCurrentUserFromPigeon(app)
    if (firebaseUser == null) {
      callback(Result.failure(FlutterFirebaseAuthPluginException.noUser()))
      return
    }

    val phoneAuthCredential = PigeonParser.getCredential(input) as? PhoneAuthCredential
    if (phoneAuthCredential == null) {
      callback(Result.failure(FlutterFirebaseAuthPluginException.invalidCredential()))
      return
    }

    reloadAfterUserUpdate(firebaseUser, callback) { it.updatePhoneNumber(phoneAuthCredential) }
  }

  override fun updateProfile(
      app: AuthPigeonFirebaseApp,
      profile: InternalUserProfile,
      callback: (Result<InternalUserDetails>) -> Unit
  ) {
    val firebaseUser = getCurrentUserFromPigeon(app)
    if (firebaseUser == null) {
      callback(Result.failure(FlutterFirebaseAuthPluginException.noUser()))
      return
    }

    val builder = UserProfileChangeRequest.Builder()
    if (profile.displayNameChanged) {
      builder.setDisplayName(profile.displayName)
    }
    if (profile.photoUrlChanged) {
      builder.setPhotoUri(profile.photoUrl?.let { Uri.parse(it) })
    }

    reloadAfterUserUpdate(firebaseUser, callback) { it.updateProfile(builder.build()) }
  }

  override fun verifyBeforeUpdateEmail(
      app: AuthPigeonFirebaseApp,
      newEmail: String,
      actionCodeSettings: InternalActionCodeSettings?,
      callback: (Result<Unit>) -> Unit
  ) {
    val firebaseUser = getCurrentUserFromPigeon(app)
    if (firebaseUser == null) {
      callback(Result.failure(FlutterFirebaseAuthPluginException.noUser()))
      return
    }

    val task =
        if (actionCodeSettings == null) {
          firebaseUser.verifyBeforeUpdateEmail(newEmail)
        } else {
          firebaseUser.verifyBeforeUpdateEmail(
              newEmail, PigeonParser.getActionCodeSettings(actionCodeSettings))
        }

    task.addOnCompleteListener { completed ->
      if (completed.isSuccessful) {
        callback(Result.success(Unit))
      } else {
        callback(
            Result.failure(
                FlutterFirebaseAuthPluginException.parserExceptionToFlutter(completed.exception)))
      }
    }
  }

  private fun reloadAfterUserUpdate(
      firebaseUser: FirebaseUser?,
      callback: (Result<InternalUserDetails>) -> Unit,
      update: (FirebaseUser) -> com.google.android.gms.tasks.Task<Void>
  ) {
    if (firebaseUser == null) {
      callback(Result.failure(FlutterFirebaseAuthPluginException.noUser()))
      return
    }

    update(firebaseUser).addOnCompleteListener { task ->
      if (task.isSuccessful) {
        firebaseUser.reload().addOnCompleteListener { reloadTask ->
          if (reloadTask.isSuccessful) {
            callback(Result.success(PigeonParser.parseFirebaseUser(firebaseUser)!!))
          } else {
            callback(
                Result.failure(
                    FlutterFirebaseAuthPluginException.parserExceptionToFlutter(
                        reloadTask.exception)))
          }
        }
      } else {
        callback(
            Result.failure(
                FlutterFirebaseAuthPluginException.parserExceptionToFlutter(task.exception)))
      }
    }
  }

  companion object {
    fun getCurrentUserFromPigeon(pigeonApp: AuthPigeonFirebaseApp): FirebaseUser? {
      val app = FirebaseApp.getInstance(pigeonApp.appName)
      val auth = FirebaseAuth.getInstance(app)
      pigeonApp.tenantId?.let { auth.setTenantId(it) }
      return auth.currentUser
    }

    fun buildOAuthProvider(signInProvider: InternalSignInProvider): OAuthProvider {
      val provider = OAuthProvider.newBuilder(signInProvider.providerId)
      signInProvider.scopes?.filterNotNull()?.let { provider.setScopes(it) }
      signInProvider.customParameters?.let { params ->
        val converted = HashMap<String, String>()
        for ((key, value) in params) {
          if (key != null && value != null) {
            converted[key] = value
          }
        }
        provider.addCustomParameters(converted)
      }
      return provider.build()
    }
  }
}
